packages/flutter_tools/lib/src/dart/package_map.dart

```diff
@@ -16,6 +16,44 @@ Future<PackageConfig> currentPackageConfig() async {
   return loadPackageConfigUri(Isolate.packageConfigSync!);
 }
 
+/// Locates the `.dart_tool/package_config.json` relevant to [dir].
+///
+/// Searches [dir] and all parent directories.
+///
+/// Returns `null` if no package_config.json was found.
+File? findPackageConfigFile(Directory dir) {
+  final FileSystem fileSystem = dir.fileSystem;
+
+  String candidateDir = fileSystem.path.absolute(dir.path);
+  while (true) {
+    final File candidatePackageConfigFile = fileSystem.file(
+      fileSystem.path.join(candidateDir, '.dart_tool', 'package_config.json'),
+    );
+    if (fileSystem.file(candidatePackageConfigFile).existsSync()) {
+      return candidatePackageConfigFile;
+    }
+    // TODO(sigurdm): we should not need to check this file, it is obsolete.
+    final File candidatePackagesFile = fileSystem.file(fileSystem.path.join(candidateDir, '.packages'));
+    if (fileSystem.file(candidatePackagesFile).existsSync()) {
+      return candidatePackagesFile;
+    }
+    final String nextDir = fileSystem.path.dirname(candidateDir);
+    if (nextDir == candidateDir) {
+      return null;
+    }
+    candidateDir = nextDir;
+  }
```

Comment:
Nit: I think this would be slightly more readable.

```suggestion
    final String parentDir = fileSystem.path.dirname(candidateDir);
    
    if (fileSystem.identicalSync(parentDir, candidateDir)) {
      return null;
    }
    candidateDir = parentDir;
```